neo/Network/P2P/Payloads/Witness.cs

```diff
@@ -1,13 +1,14 @@
 ﻿using Neo.IO;
 using Neo.IO.Json;
 using Neo.SmartContract;
-using Neo.VM;
+using System;
 using System.IO;
 
 namespace Neo.Network.P2P.Payloads
 {
     public class Witness : ISerializable
     {
+        public WitnessScope Scope;
```

Comment:
Does `Witness` need a scope? I think it is only used in `Transaction`.

Reply:
I thought a lot about it, and in the end, I believe it's responsibility of the Witness, to know its validity.
Do you think this helps, moving this to Transaction? I thought that pehaps it would become more complicated... I'll try again.

Reply:
I think we can move it to `TransactionAttribute`.

Reply:
The advantage of putting on Transaction is that we can remove these other constraints here on BlockBase, etc:
```cs
            Witness = reader.ReadSerializable<Witness>();
            if (Witness.Scope.Type != WitnessScopeType.Global) throw new FormatException();
```
is it only for Application trigger perhaps? :thinking:

Reply:
Erik, let's think...
On `Runtime_CheckWitness` we are taking witnesses from `ScriptContainer`, which is `IVerifiable`: `foreach (var witness in engine.ScriptContainer.Witnesses)`.
I wanted to do `engine.ScriptContainer as Transaction`, but it's not true, as ApplicationEngine is being used for other IVerifiable things (such as blocks, etc). This is the challenge.
We can take it out of Witness, but it would still be on IVerifiable... perhaps as an external array, what wouldn't really solve the issue for BlockBase and others. The point is: can we make sure that `CheckWitness` will only be executed for `Transaction` IVerifiable ScriptContainer?

Reply:
Yes. It is only used in `Transaction` (Application trigger).

Reply:
if is moved to an attribute, we need to allow multiple scopes, related to the signature, not just one for all of them. Also we can extend Witness, Scope Witness and use this class in Transactions

Reply:
`Scope` is an enum, isn't it? You can combine them as flags.

Reply:
This doesn't seem to be the case... looks like Blocks are IVerifiable, signed by multisig consensus for verification... meaning that we would need to know, how to process consensus signatures (in which scope). On practice, this only matters for security over Application, since verification cannot change balances (or can it? :thinking:)
One thing we can do is to default scope Global during Verification trigger, and only for AplicationTrigger (hopefully only Transaction type will be there), then we process WitnessScope (from Usage). Do you think this can work?

Reply:
In my opinion, Witness has no scope, and Cosigner has scope.

Reply:
Ok, I think that's easy to fix. IVerifiable has two main inherited componentes: Blocks and Transactions. Blocks only have one Witness, so we can put it global (a ledger decision). The rest we can get from a Transaction cast, directly from cosigners. This simplifies a lot.

Reply:
> Scope is an enum, isn't it? You can combine them as flags.

Perhaps this makes sense. Global should be a superset of all (0x07). 
RootAccess 0x03 can be combined with Custom 0x01, enforcing that scripthash is indeed the entryscripthash.
For groups we need to see, if we can merge with custom, with different Data size ( must verify how group hashes or pubkeys are working). If manifest ensures that group cannot be exceeded, when informed, then global would work for it.

Reply:
@erikzhang when a contract is part of a group,can it call something outside this group?